Sy step1

izayaki22 opened the pull request.

izayaki22 committed:
feat: 로또 구입 금액을 입력 받는 기능 추가
- Scanner를 이용하여 사용자가 직접 구입 금액을 입력하도록 함
- 이후 지난 주 당첨 번호도 입력 받아야되기에,  nextLine을 통해 입력받은 로또 구입 금액 뒤에 붙어있는 "\n"을 지움

izayaki22 committed:
feat: 로또 개수를 계산하는 기능 추가

izayaki22 committed:
feat: 구매한 로또 개수만큼 로또를 랜덤으로 생성하는 기능 추가
- makeLotto 메서드는 로또 개수를 입력받고, 생성한 로또를 2차원의 ArrayList로 만들어 반환
- 로또 번호는 1부터 45까지 중복없는 6개의 숫자를 랜덤으로 고름
- 하나의 로또를 생성하면 해당 로또를 화면에 출력함

izayaki22 committed:
feat: 지난 주 당첨 번호를 입력 받는 기능 추가
- 앞에서 생성한 Scanner를 이용
- 6개의 숫자가 ","와 함께 입력 받으므로, int가 아닌 String 타입으로 변수를 설정함

izayaki22 committed:
feat: String으로 입력 받은 지난 주 당첨 번호를 ArrayList로 바꾸는 기능 추가
- winNumbers 메서드는 사용자가 작성한 당첨 번호를 String으로 입력 받고, 번호만 ArrayList에 추가해 해당 ArrayList를 반환
- 당첨 번호를 입력 받을 때 "1,2"도 가능하고 "1, 2"도 가능하도록 delim를 ","와 " " 둘 다 사용
-

izayaki22 committed:
feat: 각각의 로또가 맞은 개수를 하나의 ArrayList로 만드는 기능 추가
- countResults 메서드는 지난 주 당첨 번호, 구매한 로또번호들, 로또의 개수를 입력받음
- 각 로또마다 당첨 번호와 몇 개가 일치하는지 개수를 센 후, ArrayList에 추가함

izayaki22 committed:
feat: 하나의 로또에 몇 개의 당첨 번호가 포함되어 있는지 개수를 세는 기능 추가
- countMathcedNumbers 메서드는 당첨 번호, 하나의 로또 번호를 입력 받고 당첨된 개수를 반환

izayaki22 committed:
feat: 로또 당첨 통계를 화면에 출력 기능 추가
- Collection 클래스의 frequency 메서드를 사용해 3개, 4개, 5개, 6개가 일치하는 로또의 개수가 각각 몇 개인지 보여줌

izayaki22 committed:
feat: 총 수익률을 계산하고 출력하는 기능 추가
- eariningRate 메서드는 당첨 개수가 저장된 ArrayList와 구매 금액을 입력 받고, 수익률을 계산하여 반환
- 수익률은 로또 당첨 금액에서 구입 금액을 뺀 후, 구입 금액을 나누고 100을 곱하여 계산
- 수익률은 소수점 둘째 자리까지만 화면에 출력

izayaki22 committed:
test: makeLotto 메서드의 테스트코드 작성
- 랜덤으로 만들어지는 로또 번호의 숫자 범위가 1 이상 45 이하인지 확인

izayaki22 committed:
fix: makeLotto 메서드의 랜덤 숫자 범위 수정
- makeLotto의 테스트 코드를 1 이상 44 이하의 범위로 돌려봤을 때 테스트가 통과되어 버그 발견
- random 인스턴스에서 nextInt 메서드의 인수를 44에서 45로 변경함

izayaki22 committed:
fix: 메서드들의 접근 범위 수정
- 테스트코드를 만들기 위해 다른 클래스에서도 해당 메서드에 접근이 가능하도록 private을 public으로 수정함

izayaki22 committed:
test: winNumbers 메서드의 테스트 코드 추가
- ","로 연결된 String을 인수로 집어넣어 winNumbers 메서드로 반환된 ArrayList와 1부터 6까지의 숫자를 직접 집어 넣은 ArrayList가 동일한지 비교하는 테스트

izayaki22 committed:
test: winNumbers 메서드의 테스트 코드 추가
- ","로 연결된 String을 잘라 넣은 ArrayList와 직접 1부터 6까지의 숫자를 집어 넣은 ArrayList가 동일한지 비교하는 테스트

izayaki22 committed:
refactor: 2차원 ArrayList의 생성 방법 변경
- 테스트코드를 만들 때 {{1, 2}, {3, 4}}와 같은 2차원 ArrayList를 Arrays 클래스의 asList 메서드를 사용해 초기화 하고싶어, ArrayList<Integer>[]에서 ArrayList<ArrayList<Integer>>로 수정

izayaki22 committed:
refactor: 2차원 ArrayList의 생성 방법 변경
- 테스트코드를 만들 때 {{1, 2}, {3, 4}}와 같은 2차원 ArrayList를 Arrays 클래스의 asList 메서드를 사용해 초기화 하고싶어, ArrayList<Integer>[]에서 ArrayList<ArrayList<Integer>>로 수정

izayaki22 committed:
test: countResults 메서드의 테스트 코드 추가
- 임의로 정한 당첨 번호와 두 개의 로또번호를 인수로 입력하여 countResults 메서드로 반환된 ArrayList와, 직접 일치하는 번호의 개수를 세어 만든 ArrayList가 동일한지 비교하는 테스트

izayaki22 committed:
Merge remote-tracking branch 'origin/SY_Step1' into SY_Step1

# Conflicts:
#	src/test/java/com/poolc/javapractice/lotto/Step1Test.java

izayaki22 committed:
test: countResults 메서드의 테스트 코드 추가
- 임의로 정한 당첨 번호와 두 개의 로또번호를 인수로 입력하여 countResults 메서드로 반환된 ArrayList와, 직접 일치하는 번호의 개수를 세어 만든 ArrayList가 동일한지 비교하는 테스트

izayaki22 committed:
test: earningRate 메서드의 테스트 코드 추가
- 로또 14000원어치를 구매하고 5000원이 당첨 되었을 때의 수익률이 -64.28와 같은지 비교하는 테스트

izayaki22 committed:
fix: 수익률이 소수점 셋째자리에서 내림 되지 않고 반올림 되던 문제 수정
- 테스트 코드를 돌릴 때 "%.2f"는 소수점 둘째자리까지 반올림을 하여 출력하므로 -64.28이 아니라 -64.29가 나오는 문제 발생
- DecimalFormat 클래스의 setRoundingMode 메서드를 통해 둘째자리까지 내림하여 반환하도록 earningRate 메서드를 수정

VSFe commented on src/main/java/com/poolc/javapractice/lotto/Step1.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.*;
+
+public class Step1 {
+    public static void main(String[] args){
```
일반적으로, 입력과 출력을 받는 메인 함수는 별도의 클래스로 분리하는게 좋습니다.

비즈니스 로직 (실제 논리적인 연산이 진행되는 부분)과 입출력 클래스를 분리해서 작성해 주세요.

VSFe commented on src/main/java/com/poolc/javapractice/lotto/Step1.java:
```diff
@@ -0,0 +1,96 @@
+                }
+            }
+
+            System.out.println(lotto.get(i).toString());
```
비즈니스 로직에는 입출력을 넣지 않는 것이 좋습니다.

VSFe commented on src/main/java/com/poolc/javapractice/lotto/Step1.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public static ArrayList<Integer> winNumbers(String winNumbers) {
+        StringTokenizer tzk = new StringTokenizer(winNumbers, ", ");
```
String을 분할하기 위해 StringTokenizer를 사용하는 것도 좋지만, 리스트화 시키려면 String.split() 메서드를 활용하는 것이 더 좋습니다.

VSFe commented on src/main/java/com/poolc/javapractice/lotto/Step1.java:
```diff
@@ -0,0 +1,96 @@
+
+    public static int countMatchedNumbers(ArrayList<Integer> winList, ArrayList<Integer> lotto){
+        int matchedCount = 0;
+        for (int i = 0; i < winList.size(); i++) {
```
ArrayList 를 사용한다면, Stream API를 사용하는 것이 더 보기 좋을 수도 있어요.

```java
public static long countMatchedNumbers(ArrayList<Integer> winList, ArrayList<Integer> lotto){
	return lotto.stream()
		.filter(winList::contains)
		.count();
}
```

- lotto에 들어 있는 숫자들을 순회하면서, 필터링을 수행합니다.
- 필터링의 기준은 각 숫자가 winList에 포함되는지 여부입니다.
- 필터링이 끝나면, 남은 숫자들의 개수를 반환합니다.

모던 자바의 핵심 중 하나가 바로 이 Stream API인데, 다른 언어 (ex. Python) 에서도 이러한 함수형 프로그래밍 매커니즘을 적극 활용하고 있는 만큼, 연습해 보시면 좋을 것 같습니다.

VSFe commented on src/main/java/com/poolc/javapractice/lotto/Step1.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public static double earningRate(ArrayList<Integer> lottoResults, int buyAmount) {
+        double totalMoney = 5000 * Collections.frequency(lottoResults, 3) + 50000 * Collections.frequency(lottoResults, 4)
```
Collections.frequency를 main 에서도 한 번, 여기에서도 한 번 계산하고 있네요.
한 번만 계산할 방법은 없을까요?

VSFe reviewed (changes requested):
수고하셨습니다! 처음이라 되게 고민이 많았을텐데, 간단하게 리뷰 남겨드렸으니 확인 후 반영 부탁드립니다~

izayaki22 committed:
refactor: 비즈니스 로직과 입력 클래스 분리
- Scanner를 사용해 사용자에게 input을 받는 부분을 Input class에 따로 만들어 main과 분리
- 구입금액을 입력하는 getBuyAmount 메서드와 지난 주 당첨 번호를 입력하는 getWinNumbers 메서드로 분리

izayaki22 committed:
refactor: 비즈니스 로직과 출력 클래스 분리
- 화면에 출력 하는 부분을 Output class에 따로 만들어 main과 분리
- 구입한 로또의 개수를 출력하는 printLottoCount 메서드, 당첨 통계를 출력하는 printWinStatistics 메서드, 수익률을 출력하는 printEarningRate 메서드로 구분

izayaki22 committed:
refactor: String을 잘라 리스트화하는 방법 변경
- StringTokenizer를 사용하는 대신 String을 delimiter로 잘라 list로 반환하는 메서드인 split() 사용

izayaki22 committed:
refactor: 몇 개의 로또 번호가 일치하는지 구하는 countMatchedNumbers 메서드 변경
- 직접 winList의 원소를 하나씩 확인하는 방법은 시간이 오래 걸림
- 하나의 로또를 stream()으로 변환하고 filter 메서드를 통해 로또에 포함된 숫자만 추출한 후 count 메서드를 통해 개수를 셈

izayaki22 committed:
feat: 3, 4, 5, 6개의 숫자가 맞은 로또의 개수를 ArrayList로 반환하는 메서드 추가
- Collection.frequency 메서드 결과를 화면에 출력할 때 한 번, 수익률을 계산할 때 또 한 번 불러오므로 중복되는 과정 삭제
- result 메서드는 각 로또마다 맞은 개수가 저장된 ArrayList를 입력 받고, 그 중에서 3개, 4개, 5개, 6개의 개수만을 따로 저장한 ArrayList를 만들어 반환

izayaki22 committed:
test: result 메서드의 테스트 코드 추가
- 7개의 로또가 각각 {1, 2, 3, 3, 3, 4, 4}개 맞았다고 가정할 때, 해당 ArrayList를 result 메서드의 인수로 입력하여 반환된 ArrayList와, 직접 3개, 4개, 5개, 6개가 일치하는 로또 개수를 세어 만든 ArrayList가 동일한지 비교하는 테스트

izayaki22 committed:
refactor: 당첨 통계를 출력할 때 3개, 4개, 5개, 6개가 일치하는 로또의 개수를 불러오는 방법 수정
- Collections.frequency 메서드가 중복으로 사용 되는 것을 줄이고자 result 메서드로 반환된 resultList의 원소를 앞에서부터 하나씩 가져오는 방법으로 수정
